src,lib: reducing C++ calls on esm legacy main resolve

H4ad opened the pull request:
<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Trott committed:
tools: refloat 7 Node.js patches to cpplint.py

Cherry-pick 12c8b4d15471cb6211b39c3a2ca5b10fa4b9f12b
Original commit message:
    This commit is a suggestion for adding a rule for NULL usages in the
    code base. This will currently report a number of errors which could be
    ignored using // NOLINT (readability/null_usage)

    PR-URL: https://github.com/nodejs/node/pull/17373
    Reviewed-By: Jon Moss <me@jonathanmoss.me>
    Reviewed-By: Anna Henningsen <anna@addaleax.net>
    Reviewed-By: Timothy Gu <timothygu99@gmail.com>
    Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
    Reviewed-By: Michael Dawson <michael_dawson@ca.ibm.com>
    Reviewed-By: Sakthipriyan Vairamani <thechargingvolcano@gmail.com>
    Reviewed-By: Tobias Nießen <tniessen@tnie.de>

Refs: https://github.com/nodejs/node/commit/12c8b4d15471cb6211b39c3a2ca5b10fa4b9f12b

Cherry-pick fc81e801913de3e3f3c0c8e26c105f983a74e539
Original commit message:

    Update cpplint.py to check for inline headers when the corresponding
    header is already included.

    PR-URL: https://github.com/nodejs/node/pull/21521
    Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
    Reviewed-By: James M Snell <jasnell@gmail.com>

Refs: https://github.com/nodejs/node/commit/fc81e801913de3e3f3c0c8e26c105f983a74e539

Cherry-pick cbc3dd997eb90d629d1b9912b7a5a40eb82343df
Original commit message:

    src, tools: add check for left leaning pointers

    This commit adds a rule to cpplint to check that pointers in the code
    base lean to the left and not right, and also fixes the violations
    reported.

    PR-URL: https://github.com/nodejs/node/pull/21010
    Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
    Reviewed-By: Anna Henningsen <anna@addaleax.net>
    Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
    Reviewed-By: James M Snell <jasnell@gmail.com>

Refs: https://github.com/nodejs/node/commit/cbc3dd997eb90d629d1b9912b7a5a40eb82343df

Cherry-pick 902998190a55d6915b881936f6dd5b6e9cca6ad8
Original commit message:

    tools: fix cpplint.py header rules

    THIS COMMIT SHOULD GO WITH THE NEXT. IT WILL FIND NEW LINT.

    PR-URL: https://github.com/nodejs/node/pull/26306
    Reviewed-By: Gireesh Punathil <gpunathi@in.ibm.com>

Refs: https://github.com/nodejs/node/commit/902998190a55d6915b881936f6dd5b6e9cca6ad8

Cherry-pick 0a25ace9c35b62ece4d32fd90b326d8063265109
Original commit message:

    tools: move cpplint configuration to .cpplint

    PR-URL: https://github.com/nodejs/node/pull/27098
    Reviewed-By: Joyee Cheung <joyeec9h3@gmail.com>
    Reviewed-By: Daniel Bevenius <daniel.bevenius@gmail.com>

Refs: https://github.com/nodejs/node/commit/0a25ace9c35b62ece4d32fd90b326d8063265109

Cherry-pick afa9a7206c26a29a2af226696c145c924a6d3754
Original commit message:

    tools: refloat update link to google styleguide for cpplint

    This commit updates two old links to Google's C++ styleguide which
    currently result in a 404 when accessed.

    PR-URL: https://github.com/nodejs/node/pull/30876
    Reviewed-By: Michaël Zasso <targos@protonmail.com>
    Reviewed-By: David Carlier <devnexen@gmail.com>
    Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
    Reviewed-By: Richard Lau <riclau@uk.ibm.com>
    Reviewed-By: Rich Trott <rtrott@gmail.com>

Refs: https://github.com/nodejs/node/commit/afa9a7206c26a29a2af226696c145c924a6d3754

Cherry-pick e23bf8f771aa0bd60e25ff079985fc29b5846403
Original commit message:

    tools,src: refloat forbid usage of v8::Persistent

    `v8::Persistent` comes with the surprising catch that it requires
    manual cleanup. `v8::Global` doesn’t, making it easier to use,
    and additionally provides move semantics. New code should always
    use `v8::Global`.

    PR-URL: https://github.com/nodejs/node/pull/31018
    Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
    Reviewed-By: Richard Lau <riclau@uk.ibm.com>
    Reviewed-By: James M Snell <jasnell@gmail.com>
    Reviewed-By: David Carlier <devnexen@gmail.com>
    Reviewed-By: Rich Trott <rtrott@gmail.com>
    Reviewed-By: Gus Caplan <me@gus.host>
    Reviewed-By: Joyee Cheung <joyeec9h3@gmail.com>
    Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
    Reviewed-By: Stephen Belanger <admin@stephenbelanger.com>

PR-URL: https://github.com/nodejs/node/pull/35569
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Daijiro Wachi <daijiro.wachi@gmail.com>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/35719
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/35866

PR-URL: https://github.com/nodejs/node/pull/36213
Reviewed-By: Franziska Hinkelmann <franziska.hinkelmann@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/36235
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/36324
Reviewed-By: Beth Griggs <bgriggs@redhat.com>

PR-URL: https://github.com/nodejs/node/pull/38851
Reviewed-By: Khaidi Chu <i@2333.moe>

PR-URL: https://github.com/nodejs/node/pull/42416
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
PR-URL: https://github.com/nodejs/node/pull/48098
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

joyeecheung committed:
tools: port js2c.py to C++

This makes it easier to use third-party dependencies in this tool
(e.g. adding compression using algorithms not available in Python).
It is also much faster - locally js2c.py takes ~1.5s to generate the
output whereas this version takes ~0.1s - and consumes less memory
(~110MB v.s. 66MB).

This also modifies the js2c.py a bit to simplify the output, making
it easier to compare with one generated by the C++ version. Locally
the output from the two are identical. We'll remove js2c.py in a
subsequent commit when the C++ version is used by default.

PR-URL: https://github.com/nodejs/node/pull/46997
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

joyeecheung committed:
build: replace js2c.py with js2c.cc

PR-URL: https://github.com/nodejs/node/pull/46997
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

marco-ippolito committed:
deps: update histogram to 0.11.7

PR-URL: https://github.com/nodejs/node/pull/47742
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>

joyeecheung committed:
bootstrap: put is_building_snapshot state in IsolateData

Previously we modify the CLI options store to indicate whether the
isolate is created for building snapshot, which is a bit hacky and
makes it difficult to tell whether the snapshot is built from the
command line or through other APIs. This patch adds
is_building_snapshot to IsolateData and use this instead when we
need to know whether the isolate is created for building snapshot
(but do not care about where that request is coming from).

PR-URL: https://github.com/nodejs/node/pull/47887
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Chengzhong Wu <legendecas@gmail.com>

joyeecheung committed:
bootstrap: throw ERR_NOT_SUPPORTED_IN_SNAPSHOT in unsupported operation

This patch adds a new ERR_NOT_SUPPORTED_IN_SNAPSHOT error and throw
it in the worker constructor.

PR-URL: https://github.com/nodejs/node/pull/47887
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Chengzhong Wu <legendecas@gmail.com>

joyeecheung committed:
src: avoid prototype access in binding templates

This patch makes the binding templates ObjectTemplates, since we
don't actually need the constructor function. This also avoids
setting the properties on prototype, and instead initializes them
directly on the object template.

Previously the initialization was similar to:

```
function Binding() {}
Binding.prototype.property = ...;
module.exports = new Binding;
```

Now it's similar to:

```
module.exports = { property: ... };
```

PR-URL: https://github.com/nodejs/node/pull/47913
Reviewed-By: Chengzhong Wu <legendecas@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

jasnell committed:
quic: add additional implementation

* add QUIC_SESSION and QUIC_STREAM to AsyncWrap
* update definitions in quic/bindingdata.h
* fixup minor discrepancies in cid.h/cid.cc
* add convenience operator in struct Path
* fixup defs.h macro definitions
* fixups in quic/preferredaddress.h/cc
* fixups in src/quic/tokens.h/cc
* fixups in quic/transportparams.h/cc
* fixups in quic/tlscontext.h/cc
* add quic/streams.h/cc placeholder
* add quic session/application implementation

PR-URL: https://github.com/nodejs/node/pull/47927
Reviewed-By: Stephen Belanger <admin@stephenbelanger.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

gabrielschulhof committed:
node-api: add status napi_cannot_run_js

Add the new status in order to distinguish a state wherein an exception
is pending from one wherein the engine is unable to execute JS. We take
advantage of the new runtime add-on version reporting in order to remain
forward compatible with add-ons that do not expect the new status code.

PR-URL: https://github.com/nodejs/node/pull/47986
Reviewed-By: Chengzhong Wu <legendecas@gmail.com>
Signed-off-by: Gabriel Schulhof <gabrielschulhof@gmail.com>

santigimeno committed:
deps: upgrade to libuv 1.45.0

- linux: introduce io_uring support https://github.com/libuv/libuv/pull/3952
- src: add new metrics APIs https://github.com/libuv/libuv/pull/3749
- unix,win: give thread pool threads an 8 MB stack https://github.com/libuv/libuv/pull/3787
- win,unix: change execution order of timers https://github.com/libuv/libuv/pull/3927

Fixes: https://github.com/nodejs/node/issues/43931
Fixes: https://github.com/nodejs/node/issues/42496
Fixes: https://github.com/nodejs/node/issues/47715
Fixes: https://github.com/nodejs/node/issues/47259
Fixes: https://github.com/nodejs/node/issues/47241
PR-URL: https://github.com/nodejs/node/pull/48078
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Debadree Chatterjee <debadree333@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>

mhdawson committed:
build: add action to close stale PRs

Add action to close PRs that are over 1 year old and
have had no comments or updates in the last 6 months.

Fixes: https://github.com/nodejs/node/issues/42981

Signed-off-by: Michael Dawson <mdawson@devrus.com>

PR-URL: https://github.com/nodejs/node/pull/48051
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>

fasenderos committed:
tools: use latest upstream commit for zlib updates

Zlib rarely gets new tags or releases, so now we use the latest commit
on the upstream default branch to check if an update is available.

Refs: https://github.com/nodejs/security-wg/issues/973
PR-URL: https://github.com/nodejs/node/pull/48054
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

aduh95 committed:
doc: harmonize fenced code snippet flags

We had a few code snippets that were using a non-descriptive tag (e.g.
`console` or `text`), whereas the actual language it's using describes
it better, and improves the syntax highlighting. This commit also
removes non-necessary leading chars (e.g. `$`, `>`, or `%`) to make it
easier for readers to copy and paste to try the command themselves.

PR-URL: https://github.com/nodejs/node/pull/48082
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Paolo Insogna <paolo@cowtech.it>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

alexsch01 committed:
doc: correct line break for Windows terminals

PR-URL: https://github.com/nodejs/node/pull/48083
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Claudio Wunder <cwunder@gnome.org>

anonrig committed:
tools: update cpplint to 1.6.1

PR-URL: https://github.com/nodejs/node/pull/48098
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>

nodejs-github-bot committed:
deps,test: update postject to 1.0.0-alpha.6

PR-URL: https://github.com/nodejs/node/pull/48072
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>

aduh95 committed:
doc: fix Windows code snippet tags

Those were wrongly marked as `bash`, when the text around it was
referencing PowerShell or Command Prompt.

PR-URL: https://github.com/nodejs/node/pull/48100
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>

panva committed:
test: ignore helper files in WPTs

PR-URL: https://github.com/nodejs/node/pull/48079
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Matthew Aitken <maitken033380023@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

jrvidal committed:
doc: clarify tty.isRaw

PR-URL: https://github.com/nodejs/node/pull/48055
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

lpinca committed:
tools: fix zconf.h path

Use the `DEPS_DIR` variable to build the path instead of hardcoding it.

PR-URL: https://github.com/nodejs/node/pull/48089
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>

Trott committed:
doc: update SEA source link

The code was moved in 3803b028d so the current source link is broken
in our docs.

PR-URL: https://github.com/nodejs/node/pull/48080
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

marco-ippolito committed:
tools: add debug logs

PR-URL: https://github.com/nodejs/node/pull/48060
Reviewed-By: Ruy Adorno <ruyadorno@google.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Deokjin Kim <deokjin81.kim@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

santigimeno committed:
test: unflake test-vm-timeout-escape-nexttick

This wasn't failing on arm boxes, increase the `runInNewContext()`
timeout a bit to make sure the code it's allowed to fail.

PR-URL: https://github.com/nodejs/node/pull/48078
Fixes: https://github.com/nodejs/node/issues/43931
Fixes: https://github.com/nodejs/node/issues/42496
Fixes: https://github.com/nodejs/node/issues/47715
Fixes: https://github.com/nodejs/node/issues/47259
Fixes: https://github.com/nodejs/node/issues/47241
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Debadree Chatterjee <debadree333@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>

abmusse committed:
test: skip test-http-pipeline-flood on IBM i

PR-URL: https://github.com/nodejs/node/pull/48048
Refs: https://github.com/nodejs/node/issues/48047
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Michael Dawson <midawson@redhat.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

marco-ippolito committed:
deps: update histogram 0.11.8

PR-URL: https://github.com/nodejs/node/pull/47742
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>

Trott committed:
tools: alphabetize CODEOWNERS

Keep each list in CODEOWNERS in alphabetical order so it is easier to
find things in the longer lists.

PR-URL: https://github.com/nodejs/node/pull/48124
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Gireesh Punathil <gpunathi@in.ibm.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

tniessen committed:
doc: fix typo in crypto legacy streams API section

PR-URL: https://github.com/nodejs/node/pull/48122
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Rich Trott <rtrott@gmail.com>

nodejs-github-bot committed:
deps: update corepack to 0.18.0

PR-URL: https://github.com/nodejs/node/pull/48091
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

nodejs-github-bot committed:
deps: update ada to 2.4.2

PR-URL: https://github.com/nodejs/node/pull/48092
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

nodejs-github-bot committed:
deps: update minimatch to 9.0.1

PR-URL: https://github.com/nodejs/node/pull/48094
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

nodejs-github-bot committed:
tools: update doc to remark-parse@10.0.2

PR-URL: https://github.com/nodejs/node/pull/48095
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Rich Trott <rtrott@gmail.com>

nodejs-github-bot committed:
tools: update lint-md-dependencies

Update to remark-parse@10.0.2, remark-stringify@10.0.3, and
rollup@3.22.0.

PR-URL: https://github.com/nodejs/node/pull/48096
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>

nodejs-github-bot committed:
tools: update eslint to 8.41.0

PR-URL: https://github.com/nodejs/node/pull/48097
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

rluvaton committed:
test_runner: display dot report as wide as the terminal width

PR-URL: https://github.com/nodejs/node/pull/48038
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

HinataKah0 committed:
test_runner: delegate stderr and stdout formatting to reporter

Introduce new `TestsStream` events `test:stderr` and `test:stdout`
to delegate `stderr` and `stdout` (e.g. `console.log()`) formatting
to the reporter. And patch existing reporters to:
- Spec: output the message as it is
- TAP: stay the same with existing `test:diagnostic`

PR-URL: https://github.com/nodejs/node/pull/48045
Fixes: https://github.com/nodejs/node/issues/48011
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

MoLow committed:
test_runner: dont split lines on `test:stdout`

PR-URL: https://github.com/nodejs/node/pull/48057
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

MoLow committed:
test_runner: pass FORCE_COLOR to child process

PR-URL: https://github.com/nodejs/node/pull/48057
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

tniessen committed:
tools: fix race condition when npm installing

The update scripts that install dependencies from npm blindly assume
that `npm install` will always install the version that was previously
fetched as `dist-tags.latest`. That is not necessarily true, so instead
always install the specific version that the script thinks is current.

PR-URL: https://github.com/nodejs/node/pull/48101
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

joyeecheung committed:
src: use Blob{Des|S}erializer for SEA blobs

PR-URL: https://github.com/nodejs/node/pull/47962
Reviewed-By: Darshan Sen <raisinten@gmail.com>

MoLow committed:
test_runner: fix test deserialize edge cases

PR-URL: https://github.com/nodejs/node/pull/48106
Fixes: https://github.com/nodejs/node/issues/48103
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

nodejs-github-bot committed:
deps: update simdutf to 3.2.12

PR-URL: https://github.com/nodejs/node/pull/48118
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>

santigimeno committed:
tools: update LICENSE and license-builder.sh

A libuv `LICENSE-extra` file was added and a couple of files were
removed (stdint-msvc2008.h and pthread-fixes.c).

PR-URL: https://github.com/nodejs/node/pull/48078
Fixes: https://github.com/nodejs/node/issues/43931
Fixes: https://github.com/nodejs/node/issues/42496
Fixes: https://github.com/nodejs/node/issues/47715
Fixes: https://github.com/nodejs/node/issues/47259
Fixes: https://github.com/nodejs/node/issues/47241
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Debadree Chatterjee <debadree333@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>

atlowChemi committed:
stream: deprecate asIndexedPairs

PR-URL: https://github.com/nodejs/node/pull/48102
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Minwoo Jung <nodecorelab@gmail.com>

RaisinTen committed:
test,doc,sea: run SEA tests on ppc64

The recent Postject upgrade, https://github.com/nodejs/node/pull/48072,
included a performance improvement for the injection operation
(see https://github.com/nodejs/postject/pull/86), so now it might be
possible to run the SEA tests on the ppc64 architecture runners on
Jenkins, which was previously getting timed out.

Signed-off-by: Darshan Sen <raisinten@gmail.com>
PR-URL: https://github.com/nodejs/node/pull/48111
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

marco-ippolito committed:
tools: add security-wg as dep updaters owner

PR-URL: https://github.com/nodejs/node/pull/48113
Refs: https://github.com/nodejs/security-wg/issues/828
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

nodejs-github-bot committed:
deps: update c-ares to 1.19.1

PR-URL: https://github.com/nodejs/node/pull/48115
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

ckcr4lyf committed:
src: check node_extra_ca_certs after openssl cfg

I recently discovered that the custom NodeJS specific OpenSSL
config section in openssl.cnf would not be respected, if the
environment variable `NODE_EXTRA_CA_CERTS` was set.

This happens even if it contains an invalid value, i.e no actual
certs are read.

Someone suggested moving the checking of extra ca certs to after
the OpenSSL config is read, and this seems to work.

PR-URL: https://github.com/nodejs/node/pull/48159
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Minwoo Jung <nodecorelab@gmail.com>
Reviewed-By: Michael Dawson <midawson@redhat.com>

Stevepurpose committed:
doc: clarify mkdir() recursive behavior

PR-URL: https://github.com/nodejs/node/pull/48109
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

tniessen committed:
src: remove INT_MAX asserts in SecretKeyGenTraits

Now that CSPRNG() does not silently fail when the length exceeds
INT_MAX anymore, there is no need for the two relevant assertions
in SecretKeyGenTraits anymore.

Refs: https://github.com/nodejs/node/pull/47515
PR-URL: https://github.com/nodejs/node/pull/48053
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Filip Skokan <panva.ip@gmail.com>

tniessen committed:
doc: improve HMAC key recommendations

Add a reference to potential problems with using strings as HMAC keys.
Also advise against exceeding the underlying hash function's block size
when generating HMAC keys from a cryptographically secure source of
entropy.

Refs: https://github.com/nodejs/node/pull/48052
Refs: https://github.com/nodejs/node/pull/37248
PR-URL: https://github.com/nodejs/node/pull/48121
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Filip Skokan <panva.ip@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

tniessen committed:
doc: update outdated section on TLSv1.3-PSK

Recent OpenSSL versions support PSKs with non-SHA-256 TLSv1.3 ciphers,
e.g., TLS_AES_256_GCM_SHA384.

PR-URL: https://github.com/nodejs/node/pull/48123
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

sinkhaha committed:
lib: use existing `isWindows` variable

PR-URL: https://github.com/nodejs/node/pull/48134
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Gerhard Stöbich <deb2001-github@yahoo.de>

marco-ippolito committed:
tools: open issue when update workflow fails

PR-URL: https://github.com/nodejs/node/pull/48018
Refs: https://github.com/nodejs/security-wg/issues/973
Reviewed-By: Richard Lau <rlau@redhat.com>

Basa198 committed:
doc: update socket.remote* properties documentation

Fixes: https://github.com/nodejs/node/issues/48061
PR-URL: https://github.com/nodejs/node/pull/48139
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Paolo Insogna <paolo@cowtech.it>
Reviewed-By: Rich Trott <rtrott@gmail.com>

mhdawson committed:
doc: add Ruy to list of TSC members

Fixes: https://github.com/nodejs/TSC/issues/1388

Signed-off-by: Michael Dawson <mdawson@devrus.com>
PR-URL: https://github.com/nodejs/node/pull/48172
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

joyeecheung committed:
src: include missing header in node_sea.h

PR-URL: https://github.com/nodejs/node/pull/48152
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Michael Dawson <midawson@redhat.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

aduh95 committed:
doc: add version info on the SEA docs

PR-URL: https://github.com/nodejs/node/pull/48173
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>

daeyeon committed:
cluster: use ObjectPrototypeHasOwnProperty

Signed-off-by: Daeyeon Jeong <daeyeon.dev@gmail.com>
PR-URL: https://github.com/nodejs/node/pull/48141
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

MoLow committed:
test: fix flaky test-runner-watch-mode

PR-URL: https://github.com/nodejs/node/pull/48144
Reviewed-By: Jacob Smith <jacob@frende.me>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

fasenderos committed:
tools: log and verify sha256sum

PR-URL: https://github.com/nodejs/node/pull/48088
Refs: https://github.com/nodejs/security-wg/issues/973
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>

mcollina committed:
http: send implicit headers on HEAD with no body

If we respond to a HEAD request with a body, we ignore all writes.
However, we must still include all implicit headers.

Fixes a regressions introduced in
https://github.com/nodejs/node/pull/47732.

Signed-off-by: Matteo Collina <hello@matteocollina.com>
PR-URL: https://github.com/nodejs/node/pull/48108
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Paolo Insogna <paolo@cowtech.it>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>

ShogunPanda committed:
net: fix family autoselection SSL connection handling

PR-URL: https://github.com/nodejs/node/pull/48189
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>

richardlau committed:
deps: set `CARES_RANDOM_FILE` for c-ares

Upstream c-ares renamed `RANDOM_FILE` to `CARES_RANDOM_FILE` some
time ago in c-ares 1.17.2.

PR-URL: https://github.com/nodejs/node/pull/48156
Refs: https://github.com/c-ares/c-ares/pull/397
Reviewed-By: Michael Dawson <midawson@redhat.com>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>

dubzzz committed:
vm: properly handle defining symbol props

This PR is a follow-up of #46615.

When bumping V8 for node 18, various bugs appeared around vm.
Some have been fixed along the flow, but there is at least one
remaining and described at:
https://github.com/facebook/jest/issues/13338.

The current fix does nothing on node 20: the new bump of v8 done for
node 20 seems to fix it. But it would fix the problem in both node 18
and node 19. I confirmed it would fix node 19 by cherry-picking the
commit on v19.x-staging and launching `make -j4 jstest` on it.

PR-URL: https://github.com/nodejs/node/pull/47572
Reviewed-By: James M Snell <jasnell@gmail.com>

dholbert committed:
doc: add `auto` intrinsic height to prevent jitter/flicker

This commit addresses a scrolling/flickering issue in the HTML version
of the docs. By adding `auto` to the `contain-intrinsic-size` CSS
property, we're asking the browser to remember the last-rendered size
for the element (once it's been rendered) instead of forcing the browser
to treat it as being 1px by 5000px when it goes offscreen.

PR-URL: https://github.com/nodejs/node/pull/48195
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

Trott committed:
doc: remove broken link for keygen

Remove broken link to deprecated keygen element.

Replacing the link with something to the relevant
part of the HTML spec might be OK, but I don't think
it's necessary. People who need to know will find it
and everyone else should be discouraged from using
this.

Support for the keygen element was removed from Chrome in 2017
and removed from Firefox in 2019. We might consider deprecating
and removing support ourselves, or at least marking the API
as legacy.

Refs: https://caniuse.com/?search=keygen
PR-URL: https://github.com/nodejs/node/pull/48176
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>

fasenderos committed:
doc: fix broken link to new folder doc/contributing/maintaining

PR-URL: https://github.com/nodejs/node/pull/48205
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Deokjin Kim <deokjin81.kim@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>

devule committed:
doc: fix typo in readline completer function section

PR-URL: https://github.com/nodejs/node/pull/48188
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Zeyu "Alex" Yang <himself65@outlook.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Daijiro Wachi <daijiro.wachi@gmail.com>

kvakil committed:
test: mark test_cannot_run_js as flaky

This test has been failing occasionally since it was introduced ~5 days
ago. It was the most common failing JS test in the most recent
reliability report. Mark it as flaky.

Fixes: https://github.com/nodejs/node/issues/48180
Refs: https://github.com/nodejs/node/pull/47986
Refs: https://github.com/nodejs/reliability/issues/576
PR-URL: https://github.com/nodejs/node/pull/48181
Reviewed-By: Chengzhong Wu <legendecas@gmail.com>
Reviewed-By: Michael Dawson <midawson@redhat.com>

atlowChemi committed:
doc: add atlowChemi to triagers

PR-URL: https://github.com/nodejs/node/pull/48104
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Qingyu Deng <i@ayase-lab.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Daeyeon Jeong <daeyeon.dev@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Feng Yu <F3n67u@outlook.com>

targos committed:
tools: refactor v8_pch config

Avoid repetition by defining a new GYP target.

PR-URL: https://github.com/nodejs/node/pull/47364
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>

RaisinTen committed:
doc,vm: clarify usage of cachedData in vm.compileFunction()

Fixes: https://github.com/nodejs/node/issues/48175
Signed-off-by: Darshan Sen <raisinten@gmail.com>
PR-URL: https://github.com/nodejs/node/pull/48193
Reviewed-By: Debadree Chatterjee <debadree333@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

cjihrig committed:
test_runner: remove redundant check from coverage

The code coverage reporting logic already filters out URLs that
don't start with 'file:', so there is no need to also filter
out URLs that start with 'node:'.

PR-URL: https://github.com/nodejs/node/pull/48070
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Debadree Chatterjee <debadree333@gmail.com>

lutaok committed:
tools: order dependency jobs alphabetically

Refs: https://github.com/nodejs/security-wg/issues/973
PR-URL: https://github.com/nodejs/node/pull/48184
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Debadree Chatterjee <debadree333@gmail.com>

ovflowd committed:
doc: update codeowners with website team

PR-URL: https://github.com/nodejs/node/pull/48197
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

fasenderos committed:
tools: deps update authenticate github api request

PR-URL: https://github.com/nodejs/node/pull/48200
Fixes: https://github.com/nodejs/node/issues/48119
Refs: https://github.com/nodejs/security-wg/issues/973
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Mestery <mestery@protonmail.com>

aduh95 committed:
tools: harmonize `dep_updaters` scripts

PR-URL: https://github.com/nodejs/node/pull/48201
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>

anonrig commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -30,14 +30,14 @@ const { getOptionValue } = require('internal/options');
 const typeFlag = getOptionValue('--input-type');
 const { URL, pathToFileURL, fileURLToPath, isURL, toPathIfFileURL } = require('internal/url');
-const { canParse: URLCanParse } = internalBinding('url');
+const { canParse: URLCanParse, legacyMainResolve: URLLegacyMainResolve } = internalBinding('url');
```
I think this should be inside `node_file.cc`

anonrig commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,41 +153,58 @@ function fileExists(url) {
+
+  switch(resolvedOption) {
+    case 0:
+      guess = new URL(`./${packageConfig.main}`, packageJSONUrl);
+      break;
+    case 1:
+      guess = new URL(`./${packageConfig.main}.js`, packageJSONUrl);
+      break;
+    case 2:
+      guess = new URL(`./${packageConfig.main}.json`, packageJSONUrl);
+      break;
+    case 3:
+      guess = new URL(`./${packageConfig.main}.node`, packageJSONUrl);
+      break;
+    case 4:
+      guess = new URL(`./${packageConfig.main}.index.js`, packageJSONUrl);
+      break;
+    case 5:
+      guess = new URL(`./${packageConfig.main}.index.json`, packageJSONUrl);
+      break;
+    case 6:
+      guess = new URL(`./${packageConfig.main}.index.node`, packageJSONUrl);
+      break;
+  }
```
This can be reduced into 1 call

H4ad replied:
I thought it was faster to return `int` from C++ than returning `string` from C++, since the URL need to be parsed in anyway using `new URL`

anonrig replied:
switch/case statement has a cost

H4ad replied:
But is faster than calling C++ again to be able to emit deprecation notice when `main` is undefined, or using some `.include` or `.startWith` to check if is included inside `href`

If it's okay to remove that deprecation notice, I can remove it as well and this code will be easier to work

anonrig replied:
I don't think removing deprecation is on the table. But you might ask as a separate PR?

H4ad replied:
Yeah, I will annotate to raise a question when I open a PR to push this modification to ask about removing that warning.

anonrig commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,41 +153,58 @@ function fileExists(url) {
+
+  switch(resolvedOption) {
+    case 7:
+      guess = new URL('./index.js', packageJSONUrl);
+      break;
+    case 8:
+      guess = new URL('./index.json', packageJSONUrl);
+      break;
+    case 9:
+      guess = new URL('./index.node', packageJSONUrl);
```
Same in here.

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+  ada::url_aggregator* base_pointer = &base.value();
+
+  auto file_path = ada::parse<ada::url_aggregator>(path, base_pointer)
+    ->to_string();
```
There isn't any `to_string` method in ada?

anonrig replied:
This needs to be `get_href()`

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+  ada::result<ada::url_aggregator> base = ada::parse<ada::url_aggregator>(base_url);
+
+  if (!base) {
+    return false;
```
if it fails, it should throw. either in c++ or in js side. since this is the default behavior of `new URL`

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+
+  ada::url_aggregator* base_pointer = &base.value();
+
+  auto file_path = ada::parse<ada::url_aggregator>(path, base_pointer)
```
Can you use `std::string_view` and not use `auto`. It's hard to track without an IDE.

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
 }
 
+bool FileExist(Environment* env, const std::string_view path, const std::string_view base_url) {
+  ada::result<ada::url_aggregator> base = ada::parse<ada::url_aggregator>(base_url);
```
This can be auto

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+
+  Environment* env = Environment::GetCurrent(args);
+
+  std::string initial_path = "./" + Utf8Value(env->isolate(), args[0].As<String>()).ToString();
```
This can be `std::string_view`

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+  Environment* env = Environment::GetCurrent(args);
+
+  std::string initial_path = "./" + Utf8Value(env->isolate(), args[0].As<String>()).ToString();
+  std::string base_path = Utf8Value(env->isolate(), args[1].As<String>()).ToString();
```
if you run `ToString()` it copies. You should always use `ToStringView()` unless necessary.

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+
+  bool resolved_option;
+
+  resolved_option = FileExist(env, initial_path, base_path);
```
This can be in the same line.

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+  resolved_option = FileExist(env, initial_path, base_path);
+
+  if (resolved_option)
+    return args.GetReturnValue().Set(0);
```
I recommend having an `enum: uint32` to store the references.

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+
+  resolved_option = FileExist(env, initial_path + ".js", base_path);
+
+  if (resolved_option)
```
Why not:

```suggestion
  if (FileExist(env, initial_path + ".js", base_path))
```

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,111 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+  if (resolved_option)
+    return args.GetReturnValue().Set(1);
+
+  resolved_option = FileExist(env, initial_path + ".json", base_path);
```
You don't need to parse the URL over and over again. If you have a ada::url_aggregator instance, and just update it using `set_pathname()` it would be 8x faster. 

Info: It takes 250~ ns to parse a URL, and double that amount, if you have a second parameter (a.k.a. base). But updating is around 20-30ns.

lpinca committed:
tools: use shasum instead of sha256sum

By default, there is no sha256sum command in macOS.

PR-URL: https://github.com/nodejs/node/pull/48229
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Yongsheng Zhang <zyszys98@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

panva committed:
Revert "test: unskip negative-settimeout.any.js WPT"

This reverts commit 8244e6c35c0e9af3976511f31db2575bd2fc9d7a.

PR-URL: https://github.com/nodejs/node/pull/48182
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

lpinca committed:
test: use lower security level in s_client

With the default security level (SECLEVEL=2), the following error

```
40E72B52DB7F0000:error:0A00018A:SSL routines:tls_process_ske_dhe:dh key
too small:../deps/openssl/openssl/ssl/statem/statem_clnt.c:2100
```

is raised on on Ubuntu 22.04 on WSL2.

PR-URL: https://github.com/nodejs/node/pull/48192
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: James M Snell <jasnell@gmail.com>

sinkhaha committed:
doc: update Buffer.allocUnsafe description

PR-URL: https://github.com/nodejs/node/pull/48183
Fixes: https://github.com/nodejs/node/issues/42821
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

nodejs-github-bot committed:
deps: update ada to 2.5.0

PR-URL: https://github.com/nodejs/node/pull/48223
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Matthew Aitken <maitken033380023@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,109 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+    auto initial_pathname = file_path->get_pathname().data();
+    file_path->set_pathname(initial_pathname + std::string(".js"));
+
+    if (FileExist(env, file_path->get_pathname().data()))
```
get_pathname() returns string_view. String view does not have a cost of passing.

H4ad replied:
~I can change to string_view here but I will later need to convert to `.data` since `libuv` don't support checking if the file exists passing a `string_view`.~

Edit: I read the other comments, you already mentioned that ahahhah

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,109 @@ void BindingData::UpdateComponents(const ada::url_components& components,
                 "kURLComponentsLength should be up-to-date");
 }
 
+bool FileExist(Environment* env, const char* file_path) {
```
This should be const std::string_view

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,109 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+
+  uv_fs_t req;
+
+  int rc = uv_fs_stat(env->event_loop(), &req, file_path, nullptr);
```
Than you can do `file_path.data()`

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,109 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+      return args.GetReturnValue().Set(6);
+  }
+
+  auto base_pointer_pathname = base->get_pathname().data();
```
get_pathname() returns std::string_view. no need to convert to char*. Always use std::string_view.

H4ad replied:
In this case, I need to be converted to `char*` since I didn't find any reliable way of concatenating `string_view` with `string` or `const char*`.

If you know a way of mutating just the last part of the `pathname`, I can keep it as `string_view`, otherwise, I will need to keep this as a string and use it to concatenate with the filename everytime.

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,109 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+
+    file_path->set_pathname(initial_pathname + std::string(".json"));
+
+    if (FileExist(env, file_path->get_pathname().data()))
```
file_path->get_pathname assign this to a variable after 375. Everytime you call the getter, there is a cost.

H4ad replied:
I did that because I was not sure if `get_pathname` will return a pointer to the newest value even after I call `set_pathname`. But if I can the implementation to `string_view`, and that `string_view` keeps the latest value, this will be way faster.

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,109 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+  return rc == 0;
+}
+
+void LegacyMainResolve(const FunctionCallbackInfo<Value>& args) {
```
Can you add a comment, this function can support V8 Fast API

anonrig commented on src/node_url.cc:
```diff
@@ -317,13 +328,109 @@ void BindingData::UpdateComponents(const ada::url_components& components,
+
+  // packageConfig.main
+  if (args[1]->IsString())  {
+    file_path = ada::parse<ada::url_aggregator>(std::string("./") + *Utf8Value(env->isolate(), args[1].As<String>()), base_pointer);
```
file_path in here can be invalid. You need to check just like in line 360.

Karnav123 committed:
doc: improve the documentation of the stdio option

Fixes: https://github.com/nodejs/node/issues/47705
PR-URL: https://github.com/nodejs/node/pull/48110
Refs: https://nodejs.org/docs/latest-v20.x/api/cluster.html#clustersettings
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

lpinca committed:
test: fix zlib version regex

Add support for subrevision in the regular expression for the zlib
version.

Refs: https://github.com/madler/zlib/blob/48c3741002ac/zlib.h#L40
PR-URL: https://github.com/nodejs/node/pull/48227
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

marco-ippolito committed:
tools: automate histogram update

PR-URL: https://github.com/nodejs/node/pull/48171
Refs: https://github.com/nodejs/security-wg/issues/828
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

aduh95 committed:
esm: handle `globalPreload` hook returning a nullish value

PR-URL: https://github.com/nodejs/node/pull/48249
Fixes: https://github.com/nodejs/node/issues/48240
Reviewed-By: Geoffrey Booth <webadmin@geoffreybooth.com>
Reviewed-By: Jacob Smith <jacob@frende.me>

bnoordhuis committed:
tools: don't gitignore base64 config.h

The file is checked into git. Ignoring it causes a very non-obvious way
of breaking tarball builds:

1. Download and unpack tarball
2. Check the sources into git with `git init; git add .; git commit -a`
3. Clean the source tree with `git clean -dfx`
4. Run `./configure && make`
5. Observe build failure because config.h is missing

Fixes: https://github.com/nodejs/node/issues/47638
PR-URL: https://github.com/nodejs/node/pull/48174
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

richardlau committed:
doc: update documentation for FIPS support

When using OpenSSL 3, Node.js supports FIPS 140-2 when used with an
appropriate OpenSSL 3 provider. It is no longer necessary to rebuild
Node.js with different build time options.

Add a section on how to configure Node.js to use an OpenSSL 3 FIPS
provider to the documentation for the `crypto` module.

PR-URL: https://github.com/nodejs/node/pull/48194
Reviewed-By: Michael Dawson <midawson@redhat.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

MoLow committed:
test_runner: stop watch mode when abortSignal aborted

PR-URL: https://github.com/nodejs/node/pull/48259
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

MoLow committed:
test_runner: emit `test:watch:drained` event

PR-URL: https://github.com/nodejs/node/pull/48259
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

nodejs-github-bot committed:
deps: update zlib to 337322d

PR-URL: https://github.com/nodejs/node/pull/48218
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

clavin committed:
doc: reserve 117 for Electron 26

PR-URL: https://github.com/nodejs/node/pull/48245
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>

MoLow committed:
test_runner: fix global after hook

PR-URL: https://github.com/nodejs/node/pull/48231
Fixes: https://github.com/nodejs/node/issues/48230
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

anonrig committed:
module: reduce the number of URL initializations

PR-URL: https://github.com/nodejs/node/pull/48272
Reviewed-By: Geoffrey Booth <webadmin@geoffreybooth.com>
Reviewed-By: Matthew Aitken <maitken033380023@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

legendecas committed:
src: add Realm document in the src README.md

PR-URL: https://github.com/nodejs/node/pull/47932
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Joyee Cheung <joyeec9h3@gmail.com>

aduh95 committed:
esm: handle more error types thrown from the loader thread

PR-URL: https://github.com/nodejs/node/pull/48247
Refs: https://github.com/nodejs/node/issues/48240
Reviewed-By: Geoffrey Booth <webadmin@geoffreybooth.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Jacob Smith <jacob@frende.me>

OttoHollmann committed:
test: adapt tests for OpenSSL 3.1

PR-URL: https://github.com/nodejs/node/pull/47859
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Richard Lau <rlau@redhat.com>

kvakil committed:
build: speed up compilation of mksnapshot output

Incremental compilation of Node.js is slow. Currently on a powerful
Linux machine, it takes about 5.8 seconds to compile
`gen/node_snapshot.cc` with g++.

As in the previous PR which dealt with `node_js2c`, we add a new build
define `NODE_MKSNAPSHOT_USE_STRING_LITERALS` which is used by
`node_mksnapshot`. When this flag is set, we emit string literals
instead of array literals for the snapshot blob and for the code cache,
i.e.:

```c++
// old: static const uint8_t X[] = { ... };
static const uint8_t *X = "...";
```

I only enabled the new flag on Linux/macOS, since those are systems that
I have available for testing. On my Linux system with gcc, it speeds up
compilation of this file by 3.7s (5.8s -> 2.1s). On my Mac system with
clang, it speeds up compilation by 1.7s (3.4s -> 1.7s).

Again, the right thing here is probably to generate separate files for
the snapshot blob and for each code cache output, but this is a nice
intermediate speedup.

Refs: https://github.com/nodejs/node/issues/47984
Refs: https://github.com/nodejs/node/pull/48160
PR-URL: https://github.com/nodejs/node/pull/48162
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Joyee Cheung <joyeec9h3@gmail.com>

richardlau committed:
doc: mark Node.js 19 as End-of-Life

PR-URL: https://github.com/nodejs/node/pull/48283
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Debadree Chatterjee <debadree333@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

TerrorJack committed:
build: set v8_enable_webassembly=false when lite mode is enabled

We should set v8_enable_webassembly=false when lite mode is enabled,
since v8_enable_lite_mode implies v8_jitless, and wasm currently
doesn't work with jitless. This is automatically handled in gn, but
seems to be not the case in gyp. Enabling lite mode without setting
v8_enable_webassembly=false leads to "Warning: disabling flag
--expose_wasm due to conflicting flags" at runtime.

PR-URL: https://github.com/nodejs/node/pull/48248
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>

indutny committed:
net: fix address iteration with autoSelectFamily

When `autoSelectFamily` is set to `true`, `net.connect` is supposed to
try connecting to both IPv4 and IPv6, interleaving the address types.
Instead, it appears that the array that holds the addresses in the order
they should be attempted was never used after being populated.

PR-URL: https://github.com/nodejs/node/pull/48258
Reviewed-By: Paolo Insogna <paolo@cowtech.it>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Juan José Arboleda <soyjuanarbol@gmail.com>

MoLow committed:
doc: document watch option for node:test run()

PR-URL: https://github.com/nodejs/node/pull/48256
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

indutny committed:
tls: reapply servername on happy eyeballs connect

When establishing a TLS connection to a server with `autoSelectFamily`
set to `true`, the `net.Socket` will call `[kWrapConnectedHandle]()` to
reinitialize the socket (in case if it got broken during previous
connect attempts). Unfortunately, prior to this patch this resulted in a
brand new `TLSWrap` instance being created for the socket. While most of
the configuration of `TLSWrap` is restored, the `servername` was sadly
dropped and not reinitalized.

With this patch `servername` will be reinitialized if there are
`tls.connect` options present on the `TLSSocket` instance, making it
possible to connect with "Happy Eyeballs" to TLS servers that require
the servername extension.

PR-URL: https://github.com/nodejs/node/pull/48255
Reviewed-By: Paolo Insogna <paolo@cowtech.it>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Minwoo Jung <nodecorelab@gmail.com>

legendecas committed:
node-api: define version 9

PR-URL: https://github.com/nodejs/node/pull/48151
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Michael Dawson <midawson@redhat.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>
Reviewed-By: Gabriel Schulhof <gabrielschulhof@gmail.com>

aduh95 committed:
lib: do not disable linter for entire files

Disabling linter for single lines is less error prone.

PR-URL: https://github.com/nodejs/node/pull/48299
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Debadree Chatterjee <debadree333@gmail.com>
Reviewed-By: Moshe Atlow <moshe@atlow.co.il>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Paolo Insogna <paolo@cowtech.it>

bengl committed:
doc: clarify array args to Buffer.from()

The code for Buffer.from() treats non-Buffer and non-Uint8Array
Array-likes as Arrays. This creates some confusion when passing various
TypedArrays to Buffer.from(). The documentation now reflects the actual
behavior.

Fixes: https://github.com/nodejs/node/issues/28725
PR-URL: https://github.com/nodejs/node/pull/48274
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: LiviaMedeiros <livia@cirno.name>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Harshitha K P <harshitha014@gmail.com>

marco-ippolito committed:
Revert "tools: open issue when update workflow fails"

This reverts commit c488558c15eb088a4b81a16fc34e5416c1d6e8d5.

PR-URL: https://github.com/nodejs/node/pull/48312
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,272 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+                            std::string* result_file_path) {
+  std::string pathname_escaped_slash;
+
+  for (std::string::size_type i = 0; (i + 2) < pathname.size(); ++i) {
```
This can be size_t

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,272 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+bool FileURLToPathWin32Impl(Environment* env,
+                            std::string_view hostname,
+                            std::string_view pathname,
+                            std::string* result_file_path) {
```
Why do return the pointer but not the reference in here?

H4ad replied:
I didn't understand what you mean here.

Are you asking why I'm returning bool instead of std::string?

anonrig replied:
std::string&

H4ad replied:
I don't know enough about C++ to know the difference between both syntaxes.

I just want to change the value of the variable I pass like `func(object)` and `object.result = 'something'`;

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,272 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  char sep = decoded_pathname[2];
+
+  if (letter < 'a' || letter > 'z' ||  // a..z A..Z
+      (sep != ':')) {
```
this extra paranthesis seems unnecessary

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,272 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+    if (pathname[i + 1] != '2') continue;
+
+    if ((pathname[i + 2] | 0x20) != 102) continue;
```
Please merge this if statement with the previous one

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,272 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+bool FileURLToPathImpl(Environment* env,
+                       ada::url_aggregator* file_url,
+                       std::string* result_file_path) {
+  if (file_url->get_protocol() != "file:") {
```
string comparisons are slow. you can get the components->scheme and compare number to number.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,272 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  }
+
+#if _WIN_32
+  return FileURLToPathWin32Impl(env,
```
You dont need 2 different functions. Just wrap them between macros.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,272 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+    THROW_IF_INSUFFICIENT_PERMISSIONS(
+        env, permission::PermissionScope::kFileSystemRead, file_path);
+
+    if (UNSAFE_FileExist(env, file_path)) return args.GetReturnValue().Set(1);
```
can you put these into a for loop?

anonrig commented on src/node_url.cc:
```diff
@@ -1,17 +1,26 @@
 #include <numeric>
 
+#if defined(__MINGW32__) || defined(_MSC_VER)
+# include <io.h>
```
We can remove the changes in this folder

dependabot[bot] committed:
meta: bump codecov/codecov-action from 3.1.3 to 3.1.4

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 3.1.3 to 3.1.4.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/894ff025c7b54547a9a2a1e9f228beae737ad3c2...eaaf4bedf32dbdc6b720b63067d99c4d77d6047d)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
PR-URL: https://github.com/nodejs/node/pull/48285
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Yagiz Nizipli <yagiz@nizipli.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

dependabot[bot] committed:
meta: bump actions/setup-python from 4.6.0 to 4.6.1

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 4.6.0 to 4.6.1.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/57ded4d7d5e986d7296eab16560982c6dd7c923b...bd6b4b6205c4dbad673328db7b31b7fab9e241c0)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
PR-URL: https://github.com/nodejs/node/pull/48286
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

dependabot[bot] committed:
meta: bump github/codeql-action from 2.3.3 to 2.3.6

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 2.3.3 to 2.3.6.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/29b1f65c5e92e24fe6b6647da1eaabe529cec70f...83f0fe6c4988d98a455712a27f0255212bba9bd4)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
PR-URL: https://github.com/nodejs/node/pull/48287
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Mestery <mestery@protonmail.com>

H4ad committed:
src,lib: reducing C++ calls of esm legacy main resolve

Instead of many C++ calls, now we make only one C++ call
to return a enum number that represents the selected state.

H4ad commented:
Closed in favor of the PR to the actual node repo.
